### DISCUSSION TITLE
How to add hover state to controls?

### DISCUSSION THREAD

**Discussion by bitttttten** [fabricjs/fabric.js#9785] [Q&A] [answered]
I was looking to apply a hover state to the controls.

i.e.

```
class MyControl extends fabric.Control {
  isHovered = false

  mouseUpHandler(eventData: MouseEvent, transformData: fabric.Transform) {
    // cannot access 'this' here
  }

  mouseDownHandler(eventData: MouseEvent, transformData: fabric.Transform) {
    // cannot access 'this' here
  }

  public render(ctx: CanvasRenderingContext2D, left: number, top: number) {
    ctx.save()
    const isHovered = this.isHovered

    ctx.restore()
  }
}
```

The problem is, `this` does not exist in the mouseUp or mouseDown handler, although we have it inside render. Is there a way to do this at all?

**Comment by asturur** [comment]
I think the 'this' issue has been fixed in fabric master around 6 months ago. So if you are in 5.3 maybe is still broken ( can be fixed easily tho )

```
 if (mouseUpHandler) {
          pointer = this.getScenePoint(e);
          mouseUpHandler.call(control, e, transform!, pointer.x, pointer.y);
        }
```

which version are you using?

**Reply by bitttttten** [reply]
Ah okay, maybe actually I was approaching it incorrectly. I was looking for hover state too, not only customising the mouseUphandler.

So, let's say I wanted a hover state:

```
new fabric.Control({
  x: 0.5,
  y: -0.5,
  cursorStyle: 'pointer',
  mouseDownHandler(e, transformData, x, y) {
    // ..
  },
  mouseOverHandler() {}, // <<<-- this is the thing that doesn't exist
})
```

**Reply by asturur** [reply, marked as answer]
Ok so there is no such a thing, but when hovering an object you can check which control is hovered, and set a value in the control, like `'hovered': true` at that point you can change the render function of the control for example to change color over that value, and call a renderAll().

You have to render the full canvas today to update a control.

Today the hovered control key is kept at '__corner' property of the object.
The control render method receives the object as an argument

The thing you have to be careful about is calling renderAll only when you see that the hover state changed.

**Reply by bitttttten** [reply]
Okay, is it really that expensive then, to call renderAll when hover state changed? Or does this just scale with the amount of items?

**Reply by asturur** [reply]
It just scale with the amount of items.
In general you want to build an app that does not over renders.
Is not expensive to do that on hover state change, but fabric shouldn't do it just because someone would want to do it, as long as there is an api that allow you to build a behavior we shouldn't make it default or hardcoded inside
